sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala

```diff
@@ -45,42 +48,101 @@ object CommandUtils extends Logging {
       val partitions = sessionState.catalog.listPartitions(catalogTable.identifier)
-      if (spark.sessionState.conf.parallelFileListingInStatsComputation) {
-        val paths = partitions.map(x => new Path(x.storage.locationUri.get))
-        val stagingDir = sessionState.conf.getConfString("hive.exec.stagingdir", ".hive-staging")
-        val pathFilter = new PathFilter with Serializable {
-          override def accept(path: Path): Boolean = isDataPath(path, stagingDir)
+      val sizeWithDeserFactorsForPartitions =
+        if (sessionState.conf.parallelFileListingInStatsComputation) {
+          val paths = partitions.map(x => new Path(x.storage.locationUri.get))
+          val stagingDir = sessionState.conf.getConfString("hive.exec.stagingdir", ".hive-staging")
+          val pathFilter = new PathFilter with Serializable {
+            override def accept(path: Path): Boolean = isDataPath(path, stagingDir)
+          }
+          val deserFactCalcEnabled = sessionState.conf.deserFactorStatCalcEnabled
+          val hadoopConf = sessionState.newHadoopConf()
+          val fileStatusSeq = InMemoryFileIndex.bulkListLeafFiles(
+            paths, hadoopConf, pathFilter, spark, areRootPaths = true)
+          fileStatusSeq.flatMap { case (_, fileStatuses) =>
+            fileStatuses.map { fileStatus =>
+              sizeInBytesWithDeserFactor(
+                deserFactCalcEnabled,
+                hadoopConf,
+                fileStatus,
+                catalogTable.storage.serde)
+            }
+          }
+        } else {
+          partitions.map { p =>
+            calculateLocationSize(
+              sessionState,
+              catalogTable.identifier,
+              p.storage.locationUri,
+              p.storage.serde)
+          }
         }
-        val fileStatusSeq = InMemoryFileIndex.bulkListLeafFiles(
-          paths, sessionState.newHadoopConf(), pathFilter, spark, areRootPaths = true)
-        fileStatusSeq.flatMap(_._2.map(_.getLen)).sum
-      } else {
-        partitions.map { p =>
-          calculateLocationSize(sessionState, catalogTable.identifier, p.storage.locationUri)
-        }.sum
+      sumSizeWithMaxDeserializationFactor(sizeWithDeserFactorsForPartitions)
+    }
+  }
+
+  def sumSizeWithMaxDeserializationFactor(
+      sizesWithFactors: Seq[SizeInBytesWithDeserFactor]): SizeInBytesWithDeserFactor = {
+    val definedFactors = sizesWithFactors.filter(_.deserFactor.isDefined).map(_.deserFactor.get)
+    SizeInBytesWithDeserFactor(
+      sizesWithFactors.map(_.sizeInBytes).sum,
+      if (definedFactors.isEmpty) None else Some(definedFactors.max))
+  }
+
+  def sizeInBytesWithDeserFactor(
+      calcDeserFact: Boolean,
+      hadoopConf: Configuration,
+      fStatus: FileStatus,
+      serde: Option[String]): SizeInBytesWithDeserFactor = {
+    assert(fStatus.isFile)
+    val factor = if (calcDeserFact) {
+      val isOrc = serde.contains("org.apache.hadoop.hive.ql.io.orc.OrcSerde") ||
```

Comment:
nit: I wonder if it makes sense to get this value from `HiveSerDe.serdeMap("orc").serde` rather than hard coding it. Whether that's feasible depends on whether the HiveSerDe object is built when the hive profile is not used (the object is defined in a file in sql/core, so maybe it doesn't need the hive profile).

Reply:
Resolved in a9a673372eb224207cb950780fa8930a48154ab0.